File: cdn-system/api/services/email_service.go
package services

import (
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"net/smtp"
	"strings"
)

type SMTPConfig struct {
	IP         string `json:"ip"`
	Port       int    `json:"port"`
	User       string `json:"user"`
	Password   string `json:"pwd"`
	UseSSL     bool   `json:"use_ssl"`
	ProxyIP    string `json:"proxy_ip"`
	ProxyPort  int    `json:"proxy_port"`
	ProxyUser  string `json:"proxy_user"`
	ProxyPwd   string `json:"proxy_pwd"`
	ProxyState bool   `json:"proxy_state"`
}

func LoadSMTPConfig() (*SMTPConfig, error) {
	cfg, err := LoadSystemConfig()
	if err != nil {
		return nil, err
	}
	raw := strings.TrimSpace(cfg["smtp"])
	if raw == "" {
		return nil, errors.New("smtp config is empty")
	}
	var smtpCfg SMTPConfig
	if err := json.Unmarshal([]byte(raw), &smtpCfg); err != nil {
		return nil, err
	}
	if strings.TrimSpace(smtpCfg.IP) == "" || smtpCfg.Port == 0 {
		return nil, errors.New("smtp config incomplete")
	}
	return &smtpCfg, nil
}

func SendEmail(to, subject, htmlBody string) error {
	to = strings.TrimSpace(to)
	if to == "" {
		return errors.New("empty email")
	}
	cfg, err := LoadSMTPConfig()
	if err != nil {
		return err
	}
	host := strings.TrimSpace(cfg.IP)
	addr := fmt.Sprintf("%s:%d", host, cfg.Port)

	headers := make([]string, 0, 6)
	headers = append(headers, "From: "+cfg.User)
	headers = append(headers, "To: "+to)
	headers = append(headers, "Subject: "+subject)
	headers = append(headers, "MIME-Version: 1.0")
	headers = append(headers, "Content-Type: text/html; charset=UTF-8")
	message := strings.Join(headers, "\r\n") + "\r\n\r\n" + htmlBody

	var client *smtp.Client
	if cfg.UseSSL {
		conn, err := tls.Dial("tcp", addr, &tls.Config{ServerName: host})
		if err != nil {
			return err
		}
		client, err = smtp.NewClient(conn, host)
		if err != nil {
			return err
		}
	} else {
		client, err = smtp.Dial(addr)
		if err != nil {
			return err
		}
		if ok, _ := client.Extension("STARTTLS"); ok {
			_ = client.StartTLS(&tls.Config{ServerName: host})
		}
	}
	defer client.Close()

	if cfg.User != "" && cfg.Password != "" {
		auth := smtp.PlainAuth("", cfg.User, cfg.Password, host)
		if err := client.Auth(auth); err != nil {
			return err
		}
	}
	if err := client.Mail(cfg.User); err != nil {
		return err
	}
	if err := client.Rcpt(to); err != nil {
		return err
	}
	writer, err := client.Data()
	if err != nil {
		return err
	}
	if _, err := writer.Write([]byte(message)); err != nil {
		_ = writer.Close()
		return err
	}
	return writer.Close()
}

File: cnn.net/src/Cnn.Api/Services/Auth/EmailService.cs
using System.Net;
using System.Net.Mail;
using System.Text.Json;
using Cnn.Api.Services.Common;

namespace Cnn.Api.Services.Auth;

public sealed class SmtpConfig
{
    public string? Ip { get; set; }
    public int Port { get; set; }
    public string? User { get; set; }
    public string? Pwd { get; set; }
    public bool UseSsl { get; set; }
    public string? ProxyIp { get; set; }
    public int ProxyPort { get; set; }
    public string? ProxyUser { get; set; }
    public string? ProxyPwd { get; set; }
    public bool ProxyState { get; set; }
}

public interface IEmailService
{
    Task<bool> SendAsync(string to, string subject, string htmlBody, CancellationToken cancellationToken);
}

public sealed class EmailService : IEmailService
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true
    };

    private readonly ISystemConfigService _systemConfigService;

    public EmailService(ISystemConfigService systemConfigService)
    {
        _systemConfigService = systemConfigService;
    }

    public async Task<bool> SendAsync(string to, string subject, string htmlBody, CancellationToken cancellationToken)
    {
        to = to.Trim();
        if (string.IsNullOrWhiteSpace(to))
        {
            return false;
        }

        var config = await LoadConfigAsync(cancellationToken);
        if (config == null)
        {
            return false;
        }

        if (string.IsNullOrWhiteSpace(config.Ip) || config.Port <= 0)
        {
            return false;
        }

        using var message = new MailMessage
        {
            From = new MailAddress(config.User ?? string.Empty),
            Subject = subject,
            Body = htmlBody,
            IsBodyHtml = true
        };
        message.To.Add(new MailAddress(to));

        using var client = new SmtpClient(config.Ip, config.Port)
        {
            EnableSsl = config.UseSsl
        };
        if (!string.IsNullOrWhiteSpace(config.User) && !string.IsNullOrWhiteSpace(config.Pwd))
        {
            client.Credentials = new NetworkCredential(config.User, config.Pwd);
        }

        try
        {
            await client.SendMailAsync(message, cancellationToken);
            return true;
        }
        catch
        {
            return false;
        }
    }

    private async Task<SmtpConfig?> LoadConfigAsync(CancellationToken cancellationToken)
    {
        var cfg = await _systemConfigService.LoadSystemConfigAsync(cancellationToken);
        if (!cfg.TryGetValue("smtp", out var raw))
        {
            return null;
        }

        raw = raw?.Trim();
        if (string.IsNullOrWhiteSpace(raw))
        {
            return null;
        }

        try
        {
            return JsonSerializer.Deserialize<SmtpConfig>(raw, JsonOptions);
        }
        catch
        {
            return null;
        }
    }
}
